confluent_kafka/admin/__init__.py

```diff
@@ -0,0 +1,622 @@
+        try:
+            result = f.result()
+            for topic, error in result.items():
+                fut = futmap.get(topic, None)
+                if fut is None:
+                    raise RuntimeError("Topic {} not found in future-map: {}".format(topic, futmap))
+
+                if error is not None:
+                    # Topic-level exception
+                    fut.set_exception(KafkaException(error))
+                else:
+                    # Topic-level success
+                    fut.set_result(None)
+        except Exception as e:
+            # Request-level exception, raise the same for all topics
+            for topic, fut in futmap.items():
+                fut.set_exception(e)
+
+    @staticmethod
+    def _make_resource_result(f, futmap):
+        """
+        Map per-resource results to per-resource futures in futmap.
+        The result value of each (successful) future is a ConfigResource.
+        """
+        try:
+            result = f.result()
+            for resource, configs in result.items():
+                fut = futmap.get(resource, None)
+                if fut is None:
+                    raise RuntimeError("Resource {} not found in future-map: {}".format(resource, futmap))
+                if resource.error is not None:
+                    # Resource-level exception
+                    fut.set_exception(KafkaException(resource.error))
+                else:
+                    # Resource-level success
+                    # configs will be a dict for describe_configs()
+                    # and None for alter_configs()
+                    fut.set_result(configs)
+        except Exception as e:
+            # Request-level exception, raise the same for all resources
+            for resource, fut in futmap.items():
+                fut.set_exception(e)
+
+    @staticmethod
+    def _make_futures(futmap_keys, class_check, make_result_fn):
+        """
+        Create futures and a futuremap for the keys in futmap_keys,
+        and create a request-level future to be bassed to the C API.
+        """
+        futmap = {}
+        for key in futmap_keys:
+            if class_check is not None and not isinstance(key, class_check):
+                raise ValueError("Expected list of {}".format(type(class_check)))
+            futmap[key] = concurrent.futures.Future()
+            if not futmap[key].set_running_or_notify_cancel():
+                raise RuntimeError("Future was cancelled prematurely")
+
+        # Create an internal future for the entire request,
+        # this future will trigger _make_..._result() and set result/exception
+        # per topic,future in futmap.
+        f = concurrent.futures.Future()
+        f.add_done_callback(lambda f: make_result_fn(f, futmap))
+
+        if not f.set_running_or_notify_cancel():
+            raise RuntimeError("Future was cancelled prematurely")
+
+        return f, futmap
+
+    def create_topics(self, new_topics, **kwargs):
```

Comment:
The main thing I was noting about this before is that the signatures of the methods `create_topics` etc. change in the subclass. The new methods conveniently hide the old ones because of `**kwargs`, but it seems like kind of a trick (and my python's not strong enough to know if there are any downsides) and i'm a bit averse to tricks. Generally, the main disadvantage of  inheritance is that coupling is stronger - everything exposed in the base classes flow through to subclasses. Except for the hiding noted above, this would apply here. The other thing that threw me was use of the suffix `Impl` in the super class because it implies to me more of a complete implementation, and I think a bit unusual in this context.

That said, I agree an inheritance relationship would be ok here. My complaint about signature differences + exposing of adminclientimpl details would be mitigated by use of Pythons poor excuse for protected methods - maybe change `AdminClientImpl`'s `create_topics` etc. to `_create_topics_impl` etc. Though, the public interface is worse in that they are visible outside the class (though by convention ignored). Also maybe change the `Impl` suffix to `Base`. I don't particularly like that either, but it's more conventional.

Alternatively, I still think composition is better here. I'd be somewhat better externally. There'd just be one externally visible property `_impl` of type `_AdminClientImpl`. And I like the name `_AdminClientImpl` in this context.